File: apps/web/src/lib/types/tenant.ts
export type TenantStatus = "pending_approval" | "active" | "suspended";
export type PropertiesViewMode = "card" | "list";

export interface Tenant {
  id: string;
  name: string;
  slug: string;
  status: TenantStatus;
  currency: string;
  timezone: string;
  locale: string;
  default_properties_view: PropertiesViewMode;
  operating_countries: string[];
  contact_email: string;
  contact_phone: string;
  property_ref_prefix: string;
  banner_url: string | null;
  aggregator_enabled: boolean;
  aggregator_disabled_at: string | null;
  aggregator_disabled_reason: string | null;
  /** Platform master switch — only platform admins can toggle. */
  public_site_feature_enabled: boolean;
  /** Tenant's own on/off toggle. */
  public_site_enabled: boolean;
  /** When true, only listing-detail URLs resolve; index/profile/agents → 404. */
  public_site_direct_links_only: boolean;
  approved_at: string | null;
  approved_by_id: string | null;
  suspended_at: string | null;
  suspended_reason: string | null;
  // Per-tenant media caps (admin-set, read-only for tenants)
  max_images_per_property: number;
  max_videos_per_property: number;
  max_image_mb: number;
  max_video_mb: number;
  // migration 0051 — doc request default instructions (HTML string)
  document_request_default_instructions: string;
  created_at: string;
  updated_at: string;
}

export interface TenantMediaLimitsRequest {
  max_images_per_property: number;
  max_videos_per_property: number;
  max_image_mb: number;
  max_video_mb: number;
}

export interface TenantUpdateRequest {
  name?: string;
  currency?: string;
  timezone?: string;
  locale?: string;
  default_properties_view?: PropertiesViewMode;
  operating_countries?: string[];
  banner_url?: string | null;
  document_request_default_instructions?: string;
}

export interface TenantListResponse {
  items: Tenant[];
  total: number;
  page: number;
  page_size: number;
}

export interface TenantListParams {
  status?: string;
  q?: string;
  page?: number;
  page_size?: number;
}

export interface TenantSubscriptionInfo {
  plan: string | null;
  status: string | null;
  expires_at: string | null;
}

export interface TenantActivityEntry {
  id: string;
  actor_name: string | null;
  action: string;
  diff_summary: string | null;
  created_at: string;
}

export interface TenantCounts {
  users: number;
  agents: number;
  properties: number;
  listings: number;
  leads: number;
  deals: number;
}

export interface TenantDetailResponse extends Tenant {
  counts: TenantCounts;
  recent_activity: TenantActivityEntry[];
  subscription_info: TenantSubscriptionInfo | null;
  office_country: string | null;
  office_city: string | null;
  signup_notes: string | null;
}

export interface TenantLifecycleRequest {
  reason_note?: string;
}

export interface PublicSiteFeatureRequest {
  enabled: boolean;
}

export interface TenantSendMessageRequest {
  subject: string;
  body_text: string;
}

export interface PropertyRefPrefixRequest {
  property_ref_prefix: string;
}

export interface AggregatorStatusRequest {
  enabled: boolean;
  reason?: string;
}

File: apps/api/models/tenant.py
"""Tenant model — the root multi-tenancy anchor.

Tenant deliberately does NOT use TenantMixin; it IS the tenant.
approved_by_id is a soft FK to platform_users.id (no DB-level FK to avoid
circular dependency in migration ordering — the service layer enforces this).
suspended_by_platform_user_id has a DB-level FK to platform_users(id) with
ON DELETE SET NULL (added in 0018).
operating_countries is a Postgres TEXT[] column added in 0017.
"""

import uuid
from datetime import datetime

from sqlalchemy import Boolean, Enum, ForeignKey, Integer, String, Text, text
from sqlalchemy.dialects.postgresql import ARRAY, JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from packages.common.db.base import Base, TimestampMixin, UUIDMixin

from .enums import PropertiesViewMode, TenantStatus


class Tenant(Base, UUIDMixin, TimestampMixin):
    __tablename__ = "tenants"

    name: Mapped[str] = mapped_column(String(255), nullable=False)

    # URL-safe slug — unique across all tenants, used in subdomains / routes.
    slug: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
        unique=True,
        index=True,
    )

    status: Mapped[TenantStatus] = mapped_column(
        Enum(TenantStatus, name="tenant_status", create_type=False,
             values_callable=lambda x: [e.value for e in x]),
        nullable=False,
        default=TenantStatus.PENDING_APPROVAL,
        server_default=TenantStatus.PENDING_APPROVAL.value,
    )

    # Locale / regional defaults.
    currency: Mapped[str] = mapped_column(
        String(10), nullable=False, default="AED", server_default="AED"
    )
    timezone: Mapped[str] = mapped_column(
        String(64), nullable=False, default="Asia/Dubai", server_default="Asia/Dubai"
    )
    locale: Mapped[str] = mapped_column(
        String(10), nullable=False, default="en", server_default="en"
    )

    # Public company contact details — separate from the owner's personal email.
    contact_email: Mapped[str] = mapped_column(
        String(255),
        nullable=False,
        index=True,
    )
    contact_phone: Mapped[str] = mapped_column(
        String(32),
        nullable=False,
    )

    # Public site settings — two-gate visibility + branding.
    # Gate 1 (platform): platform admin master switch. When FALSE nothing public
    # works regardless of tenant setting.
    public_site_feature_enabled: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=False,
        server_default="false",
    )
    # Gate 2a (tenant): tenant-level on/off toggle.
    public_site_enabled: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=False,
        server_default="false",
    )
    # Gate 2b (tenant): fine-grain — when TRUE only listing-detail resolves.
    public_site_direct_links_only: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=False,
        server_default="false",
    )
    public_site_logo_url: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )
    public_site_tagline: Mapped[str | None] = mapped_column(
        String(200),
        nullable=True,
    )

    # UI preferences — default view mode for the properties list.
    default_properties_view: Mapped[PropertiesViewMode] = mapped_column(
        Enum(PropertiesViewMode, name="properties_view_mode", create_type=False,
             values_callable=lambda x: [e.value for e in x]),
        nullable=False,
        default=PropertiesViewMode.CARD,
        server_default=PropertiesViewMode.CARD.value,
    )

    # Per-tenant public-site customization blob. Schema validation in Pydantic.
    public_site_config: Mapped[dict] = mapped_column(
        JSONB,
        nullable=False,
        server_default=text("'{}'::jsonb"),
    )

    # ISO 3166-1 alpha-2 country codes the tenant operates in.
    operating_countries: Mapped[list] = mapped_column(
        ARRAY(Text),
        nullable=False,
        server_default=text("ARRAY['AE']::text[]"),
    )

    # Approval trail — nullable until a platform admin approves.
    approved_at: Mapped[datetime | None] = mapped_column(
        nullable=True,
    )
    # Soft reference to platform_users.id; no DB-level FK (see docstring).
    approved_by_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        nullable=True,
    )

    # Per-tenant media upload caps — platform admins can raise above the floor.
    # Floors are enforced at DB level (CHECK constraints in migration 0030) and
    # at application level (TenantMediaLimitsRequest validators).
    max_images_per_property: Mapped[int] = mapped_column(
        Integer, nullable=False, default=20, server_default="20"
    )
    max_videos_per_property: Mapped[int] = mapped_column(
        Integer, nullable=False, default=2, server_default="2"
    )
    max_image_mb: Mapped[int] = mapped_column(
        Integer, nullable=False, default=10, server_default="10"
    )
    max_video_mb: Mapped[int] = mapped_column(
        Integer, nullable=False, default=25, server_default="25"
    )

    # Suspension audit trail — populated when status transitions to SUSPENDED.
    suspended_at: Mapped[datetime | None] = mapped_column(
        nullable=True,
    )
    suspended_reason: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )
    suspended_by_platform_user_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("platform_users.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
